sdk/metrics/sdk.go

```diff
@@ -0,0 +1,386 @@
+// periodic collection by calling the `Collect()` method.
+func New(exporter export.MetricBatcher) *SDK {
+	m := &SDK{
+		pool: sync.Pool{
+			New: func() interface{} {
+				return &bytes.Buffer{}
+			},
+		},
+		exporter: exporter,
+	}
+	m.empty.meter = m
+	return m
+}
+
+// DefineLabels returns a LabelSet corresponding to the arguments.
+// Labels are de-duplicated, with last-value-wins semantics.
+func (m *SDK) DefineLabels(_ context.Context, kvs ...core.KeyValue) api.LabelSet {
+	// Note: This computes a canonical encoding of the labels to
+	// use as a map key.  It happens to use the encoding used by
+	// statsd for labels, allowing an optimization for statsd
+	// exporters.  This could be made configurable in the
+	// constructor, to support the same optimization for different
+	// exporters.
+
+	// Check for empty set.
+	if len(kvs) == 0 {
+		return &m.empty
+	}
+
+	// Sort and de-duplicate.
+	sorted := sortedLabels(kvs)
+	sort.Stable(&sorted)
+	oi := 1
+	for i := 1; i < len(sorted); i++ {
+		if sorted[i-1].Key == sorted[i].Key {
+			sorted[oi-1] = sorted[i]
+			continue
+		}
+		sorted[oi] = sorted[i]
+		oi++
+	}
+	sorted = sorted[0:oi]
+
+	// Serialize.
+	buf := m.pool.Get().(*bytes.Buffer)
+	defer m.pool.Put(buf)
+	buf.Reset()
+	_, _ = buf.WriteRune('|')
+	delimiter := '#'
+	for _, kv := range sorted {
+		_, _ = buf.WriteRune(delimiter)
+		_, _ = buf.WriteString(string(kv.Key))
+		_, _ = buf.WriteRune(':')
+		_, _ = buf.WriteString(kv.Value.Emit())
+		delimiter = ','
+	}
+
+	return &labels{
+		meter:   m,
+		sorted:  sorted,
+		encoded: buf.String(),
+	}
+}
+
+// labsFor sanitizes the input LabelSet.  The input will be rejected
+// if it was created by another Meter instance, for example.
+func (m *SDK) labsFor(ls api.LabelSet) *labels {
+	if l, _ := ls.(*labels); l != nil && l.meter == m {
+		return l
+	}
+	return &m.empty
+}
+
+// NewHandle returns an interface used to implement handles in the
+// OpenTelemetry API.  This takes a reference and will persist until
+// the recorder is deleted.
+func (m *SDK) NewHandle(inst api.ExplicitReportingMetric, ls api.LabelSet) api.Handle {
+	desc := inst.Descriptor()
+	// Create lookup key for sync.Map
+	labs := m.labsFor(ls)
+	mk := mapkey{
+		id:      desc.ID(),
+		encoded: labs.encoded,
+	}
+
+	rec := &record{
+		labels:         labs,
+		descriptor:     desc,
+		refcount:       1,
+		collectedEpoch: -1,
+		modifiedEpoch:  0,
+	}
+
+	rec.recorder = m.exporter.AggregatorFor(rec)
+
+	// Load/Store
+	if actual, loaded := m.current.LoadOrStore(mk, rec); loaded {
+		// Existing record case.
+		rec = actual.(*record)
+		atomic.AddInt64(&rec.refcount, 1)
+		return rec
+	}
+	m.addPrimary(rec)
+	return rec
+}
+
+// DeleteHandle removes a reference on the underlying record.  This
+// method checks for a race with `Collect()`, which may have (tried to)
+// reclaim an idle recorder, and restores the situation.
+func (m *SDK) DeleteHandle(r api.Handle) {
```

Comment:
I think that the exporter should be let know about the handle being deleted, it might need to do some cleanups on its own.

Reply:
The prometheus libraries support a `Delete()` on their handles, but it's not meant to be used in this way. Prometheus essentially expects you never to remove handles from an exporter, so this is a grey area.  I think exporters should take care of this problem on their own. A push exporter probably won't have to worry about it.